test/main/modules.js

```diff
@@ -1,178 +1,483 @@
 
-      if (typeof this.beforeCount === 'undefined') {
-        this.beforeCount = 0;
-      }
-
-      this.beforeCount++;
-    }
-  });
+    hooks.before(function () {
+      x.push('parent-before: ' + flat(this));
+      this.before = (this.before || '') + 'P';
+    });
+    hooks.beforeEach(function () {
+      x.push('parent-beforeEach: ' + flat(this));
+      this.beforeEach = (this.beforeEach || '') + 'P';
+    });
+    hooks.afterEach(function () {
+      x.push('parent-afterEach: ' + flat(this));
+      this.afterEach = (this.afterEach || '') + 'P';
+    });
+    hooks.after(function () {
+      x.push('parent-after: ' + flat(this));
+      this.after = (this.after || '') + 'P';
+    });
 
-  QUnit.test('runs before first test', function (assert) {
-    assert.expect(2);
-    assert.equal(this.beforeCount, 1, 'beforeCount should be one');
-  });
+    QUnit.module('child', function (hooks) {
+      hooks.before(function () {
+        x.push('child-before: ' + flat(this));
+        this.before = (this.before || '') + 'C';
+      });
+      hooks.beforeEach(function () {
+        x.push('child-beforeEach: ' + flat(this));
+        this.beforeEach = (this.beforeEach || '') + 'C';
+      });
+      hooks.afterEach(function () {
+        x.push('child-afterEach: ' + flat(this));
+        this.afterEach = (this.afterEach || '') + 'C';
+      });
+      hooks.after(function () {
+        x.push('child-after: ' + flat(this));
+        this.after = (this.after || '') + 'C';
+      });
 
-  QUnit.test('does not run before subsequent tests', function (assert) {
-    assert.expect(1);
-    assert.equal(this.beforeCount, 1, 'beforeCount did not increase from last test');
-  });
+      QUnit.test('one', function (assert) {
+        x.push('child-test: ' + flat(this));
+        this.tester = '1';
+        assert.true(true);
+      });
+    });
 
-  QUnit.module('before (skip)', {
-    before: function (assert) {
-      assert.true(true, 'before hook ran');
-    }
+    QUnit.test('two', function (assert) {
+      x.push('parent-test: ' + flat(this));
+      this.tester = '2';
+      assert.true(true);
+    });
   });
 
-  QUnit.skip('first test in module is skipped');
+  QUnit.module('parent with initial test', function (hooks) {
+    var x = actual['parent with initial test'] = [];
 
-  QUnit.test('runs before first unskipped test', function (assert) {
-    assert.expect(1);
-  });
+    hooks.before(function () {
+      x.push('parent-before: ' + flat(this));
+      this.before = (this.before || '') + 'P';
+    });
+    hooks.beforeEach(function () {
+      x.push('parent-beforeEach: ' + flat(this));
+      this.beforeEach = (this.beforeEach || '') + 'P';
+    });
+    hooks.afterEach(function () {
+      x.push('parent-afterEach: ' + flat(this));
+      this.afterEach = (this.afterEach || '') + 'P';
+    });
+    hooks.after(function () {
+      x.push('parent-after: ' + flat(this));
+      this.after = (this.after || '') + 'P';
```

Comment:
not sure this is a problem, but since we don't push onto x after the above line, nothing ever asserts that the `P` makes it onto the environment (except in the multiple-hooks case where the P2 shows up in the last after assertion)

Reply:
I think this is a "tree falls in the forrest when no-one is around" type of situation. If there's no place to observe it within the API, there's not really a distinction between whether we do what we logically expect (pass the original `this` from before as-is) vs what would be "wrong" (pass a copy that we throw away later), since this is the last time the object is seen, after which we throw away the entire module.

If this isn't the last time, I stand corrected of course :)